src/Engine/ProtoCore/Lang/CallSite.cs

```diff
@@ -1671,6 +1677,16 @@ private FunctionGroup FirstFunctionGroupInInheritanceChain(RuntimeCore runtimeCo
                 //We will call the subsequent reductions n times
                 int cartIndex = ri.CartesianIndex;
 
+                //We determine if the input paramaters are homogenious to set the final function endpoint once
+                if (cartIndex == 0)
+                {
+                    List<StackValue> finalFormalParameters = new List<StackValue>();
+                    var homogenious = AreParametersHomogenious(formalParameters, runtimeCore, finalFormalParameters);
+
+                    if(homogenious)
+                        finalFuntionEndPoint = SelectFinalFep(c, functionEndPoint, finalFormalParameters, stackFrame, runtimeCore);
+                }
+
```

Comment:
Can this block of changes go before:
```
if (replicationInstructions.Count == 0)
{
               return ExecWithZeroRI(functionEndPoint, c, formalParameters, stackFrame, runtimeCore, previousTraceData, newTraceData, finalFuntionEndPoint);
 }
```

Reply:
Let me look.

Reply:
I think not for the same reason I mentioned before.... I was assuming that we wanted to validate the function input parameters are the same types (homogeneous) with the initial call of  `ExecWithRISlowPath()` vs at every nested invocation.  That leaves some situations where you could have a partial optimization uncovered (ie they drop back to determining the Function endpoint with every call) but this may be simpler.